### ISSUE TITLE
Add a mechanism for more content types to servant-foreign

### ISSUE THREAD

**Issue by arianvp** [haskell-servant/servant#290]
As discussed in #287, it would be neat to have code generators that can also handle non-JSON API's. Currently servant-foreign only works if the API that you provide has JSON for all the  request bodies and for all response bodies. We probably want to generalize this.

**Comment by fisx** [comment]
Scratch the "probably": I am working on production APIs that give mixed JSON, PNG, and WAV responses.

**Comment by arianvp** [comment]
Would you have some ideas how you would want to use this feature? How do you currently use PNG and WAV through XHR?

**Comment by jkarni** [comment]
Would we still limit code generators to a single one of the possible content-types (e.g. the first)? Or do we also want to add the option of picking content-type (by e.g. having an extra param to the function)?

**Comment by arianvp** [comment]
I think this really depends on the underlying framework of the language
you're targeting.  I think both are possible.

**Comment by jhstanton** [comment]
It seems like this was put on the back burner. Has anyone looked into how to implement this? Even if it was just adding documentation for making instances for other content-types would be helpful. 

Currently I'm looking at doing something like this to get around the issue:

``` Haskell
type APIForJS             = "a" :> "b" :> Get '[JSON] MyType
type ProblemEndpoint = "problem" :> Get '[OctetStream] ByteString

type CompleteAPI       = APIForJS :<|> ProblemEndpoint
```

then generating the JS with just a subset of my total endpoints (and writing the rest). Not the biggest issue in the world, but it would be nice if users could help the generator along.

**Comment by alpmestan** [comment]
If anyone wants to give this a shot we can definitely try and brainstorm a solution to this =)

**Comment by exarkun** [comment]
I have no idea what the solution to this might be but I do have a use-case for non-JSON.  Specifically, OctetStream.  I don't need to think about how this interacts with XHR because I'm focused on generating a Python client.  I would love to help get this addressed.

Perhaps this is naive but it seems like Servant.Foreign doesn't need to concern itself with how Servant.JS deals with non-JSON, does it?  I wonder if the earlier comments on this issue were from a time when the two projects were one.

I would like Servant.Foreign to expose the underlying information in a way that modules like Servant.PY and Servant.JS can deal with however they choose.

**Comment by alpmestan** [comment]
Yes, servant-foreign got "extracted" out of servant-js back in the days. I'd be delighted if someone feels like doing whatever it takes to get servant-foreign to nicely support other content types. It might require a bit of design beforehand (how do you then tell foreign languages which content type this endpoint takes or returns for the request/response body, etc), but it's definitely feasible.

**Comment by domenkozar** [comment]
I've used a bit of a hack to filter out the stuff I don't need with `servant-flatten`:

```haskell
api = Proxy :: Proxy (FilterAPI (Flat API))

type family FilterAPI api where
  FilterAPI (Stream method status framing ct a) = EmptyAPI
  FilterAPI (StreamBody' mods framing ct a :> b) = EmptyAPI
  FilterAPI (Verb method statusCode '[FormUrlEncoded] a) = EmptyAPI
  FilterAPI (a :<|> b) = FilterAPI a :<|> FilterAPI b
  FilterAPI (reqBody '[FormUrlEncoded] a :> b) = EmptyAPI
  FilterAPI ((a :: k) :> b) = a :> FilterAPI b
  FilterAPI a = a
```

**Comment by georgefst** [comment]
For what it's worth, it seems there's nothing in `servant-foreign` that's actually _specific to_ JSON. I'm dealing with an API where all content types are plain text, and I've managed to use it by forking the library and literally just doing a find and replace: `find src -type f -name '*.hs' -exec sed -i 's/JSON/PlainText/g' {} \;`.

Obviously that doesn't mean that generalising the API to make it possible to deal with multiple content types at once is necessarily easy. But it seemed worth mentioning.

**Comment by georgefst** [comment]
Huh, now I've just seen #848, and I don't quite understand why it was rejected.
